File: mergeKSortedListsHeap.py
"""
Time Complexity : O(nklogk) where n is the average number of elements in the array and k is the number lists.
                  Heapify will take place for k times so logk and it will be performed on all the elements hence nk
Space Complexity : O(logk ) where k is the number of lists.

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

Your code here along with comments explaining your approach:
Idea is to perform the Heap Sort. 
1. Initially, first elements of all the list will be inserted to heap. At the time of insertion heapify will take place and we will pop the minimum element (the top element will be the smallest as all the lists are already sorted and we are perfroming heapify on minHeap).
2. After that whatever elements is pop, we will go to the next element in that array from which the element was popped and again heapify will take place.
3. This process will take place until all the elements in all the lists are visited and the heap becomes empty
"""
import heapq
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        result = ListNode(None)
        curr = result
        for i , head in enumerate(lists):
            if head:
                heapq.heappush(heap, (head.val, i, head))
    
        while(heap):
            minVal, minIdx, minNode = heapq.heappop(heap)
            minNext = minNode.next
            minNode.next = None
            curr.next = minNode
            curr = curr.next
            
            if minNext:
                heapq.heappush(heap, (minNext.val, minIdx, minNext))
        
        return result.next
            

File: MergeKSortedListsHeap.java
import java.util.PriorityQueue;

public class MergeKSortedListsHeap {


        public ListNode mergeKLists(ListNode[] lists) {

            // pq (heap) of list node values
            PriorityQueue<ListNode> pq = new PriorityQueue<>((a,b) -> a.val - b.val);

            // add head nodes of all lists to heap
            for(ListNode Head: lists) {

                if(Head != null) pq.add(Head);
            }

            // for adding the first (smallest) list node
            ListNode dummy = new ListNode(-1);

            // to move current pointer of result list
            ListNode curr = dummy;

            // as long as priority queue is not empty
            while(!pq.isEmpty()) {

                // get minimum of k size heap
                ListNode min = pq.poll();

                // add this minimum to current pointer of result list
                curr.next = min;

                // move current pointer further (recently added minimum)
                curr = curr.next;

                // if minimum has a next list node in its list, add it to heap
                if(min.next != null) pq.add(min.next);
            }
            // result starts after dummy pointer
            return dummy.next;
        }



}

/*
TIME COMPLEXITY = O(nklogk)

heap of size - k
size of each list - n

SPACE COMPLEXITY = O(k)

O(n) space - for a new linked list
*/